### PR TITLE
feat: phone authentication

### PR DESCRIPTION
**Pull request by TorstenDittmann**
## Todo
- [x] fix naming conventions for the classes (they are all over the place right now)
- [x] considering the approach for env variables we have for storage
- [x] add missing tests for the users service
- [x] add all the documentation
- [x] update UI

## What does this PR do?

Adds the option to use your phone as a method of authentication.

The initial implementation will allow using the following 3 providers:
- [Twilio](https://twilio.com/)
- [Telesign](https://www.telesign.com/)
- [TextMagic](https://www.textmagic.com/)

A Messaging Worker has been added to the Stack.

Following attributes have been added to the user:
- phone
- phoneVerification

Following env variables have been added:
- _APP_PHONE_PROVIDER
- _APP_PHONE_USER
- _APP_PHONE_SECRET
- _APP_PHONE_FROM

Following endpoints have been added:

**Create Phone Session**
`post('/v1/account/sessions/phone')`
- userId
- number

Will initialize the phone login and sends 6 random digits to the phone.

**Create Phone session (confirmation)**
`put('/v1/account/sessions/phone')`
- userId
- secret

Will confirm the the session and create a user account if not found. The 6 digits will be passed as the secret.

**Update Account Phone**
`patch('/v1/account/phone')`
- number
- password

Updates the users phone number.

**Create Phone Verification**
`post('/v1/account/verification/phone')`
- *no params*

Creates a verification token and sends 6 digits to the user's phone.

**Create Phone Verification (confirmation)**
`put('/v1/account/verification/phone')`
- userId
- secret

Confirms the phone number by passing the 6 digits as the secret.

**Update Phone**
`patch('/v1/users/:userId/email')`
- userId
- number

Updates the user's phone number.

**Update Phone Verification**
`patch('/v1/users/:userId/verification')`
- userId
- phoneVerification

Updates the phoneVerification attribute on the user.

## Test Plan

- added tests for all the new endpoints using a mock adapter for the
- manually tested all 3 adapters

## Related PRs and Issues

- #3203 
- #1068 

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

✅

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -56,6 +56,10 @@ _APP_SMTP_PORT=1025
 _APP_SMTP_SECURE=
 _APP_SMTP_USERNAME=
 _APP_SMTP_PASSWORD=
+_APP_PHONE_PROVIDER=mock
```
Comment (eldadfux) [inline .env:59]: The DSN approach that @christyjacob4 is implementing for databases might work better here and possibly for storage as well.

File: app/controllers/api/users.php [hunk 2]
```diff
@@ -438,6 +439,45 @@
...
+    ->label('scope', 'users.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'users')
+    ->label('sdk.method', 'updatePhoneVerification')
+    ->label('sdk.description', '/docs/references/users/update-user-phone-verification.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_USER)
+    ->param('userId', '', new UID(), 'User ID.')
+    ->param('phoneVerification', false, new Boolean(), 'User phone verification status.')
```
Comment (eldadfux) [inline app/controllers/api/users.php:455]: We can drop the prefix here:`phoneVerification`?

File: app/controllers/api/users.php [hunk 3]
```diff
@@ -438,6 +439,45 @@
         $response->dynamic($user, Response::MODEL_USER);
     });
 
+App::patch('/v1/users/:userId/verification')
```
Comment (eldadfux) [inline app/controllers/api/users.php:442]: Isn't this conflicting with email verification? 
```suggestion
App::patch('/v1/users/:userId/verification/phone')
```

File: app/controllers/api/users.php (hunk 2 above)
Comment (eldadfux) [reply app/controllers/api/users.php:455]: OK, I understand why you did it like this.

File: app/init.php [hunk 4]
```diff
@@ -64,6 +69,8 @@
 use Utopia\Storage\Device\Linode;
 use Utopia\Storage\Device\Wasabi;
 
+use function PHPUnit\Framework\matches;
```
Comment (eldadfux) [inline app/init.php:72]: Leftover

File: app/workers/messaging.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+use Appwrite\Auth\Phone;
+use Appwrite\Auth\Phone\{
```
Comment (eldadfux) [inline app/workers/messaging.php:4]: Let's use a consistent syntax, one way or another.

File: src/Appwrite/Auth/Phone.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * @param string $method
+     * @param string $url
+     * @param array  $headers
+     * @param string $payload
+     *
+     * @return string
+     */
+    protected function request(string $method, string $url, array $headers = [], ?string $payload = null, ?string $userpwd = null): string
```
Comment (eldadfux) [inline src/Appwrite/Auth/Phone.php:45]: We usually call this method `call`

File: src/Appwrite/Utopia/Response/Model/User.php [hunk 7]
```diff
@@ -47,12 +47,24 @@ public function __construct()
                 'default' => '',
                 'example' => 'john@appwrite.io',
             ])
+            ->addRule('phone', [
+                'type' => self::TYPE_STRING,
+                'description' => 'User phone number.',
+                'default' => '',
+                'example' => '+49 30 901820',
```
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/User.php:54]: Should we keep spaces?

File: src/Appwrite/Utopia/Response/Model/User.php [hunk 8]
```diff
@@ -47,12 +47,24 @@ public function __construct()
                 'default' => '',
                 'example' => 'john@appwrite.io',
             ])
+            ->addRule('phone', [
+                'type' => self::TYPE_STRING,
+                'description' => 'User phone number.',
```
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/User.php:52]: We should mention the format to expect with link to relevant information about it.

File: app/controllers/api/account.php [hunk 9]
```diff
@@ -519,7 +522,7 @@
             'countryCode' => ($record) ? \strtolower($record['country']['iso_code']) : '--',
```
Comment (eldadfux) [inline app/controllers/api/account.php:522]: I still think we should change the default create session endpoint to specifically create a session with email. Currently it seems like the API has a bias and suggest email is *the* way to go.

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Appwrite/Auth/Phone.php (hunk 6 above)
Comment (TorstenDittmann) [reply src/Appwrite/Auth/Phone.php:45]: I got this from OAuth, I can change both either way.

File: src/Appwrite/Utopia/Response/Model/User.php (hunk 7 above)
Comment (TorstenDittmann) [reply src/Appwrite/Utopia/Response/Model/User.php:54]: Should definitely be removed here.

File: app/controllers/api/account.php (hunk 9 above)
Comment (TorstenDittmann) [reply app/controllers/api/account.php:522]: We can do that 👍 I initially did this for less breaking change and less work for me 😉

**Pull request merged**